In [201]:
import numpy as np
import pandas as pd

In [202]:
df = pd.read_csv('Zip_Zri_MultiFamilyResidenceRental-2.csv',index_col=False)
zillow= df.copy()

In [203]:
zillow

RegionID  RegionName            City State  \
0        61639       10025        New York    NY   
1        84654       60657         Chicago    IL   
2        61637       10023        New York    NY   
3        91982       77494            Katy    TX   
4        84616       60614         Chicago    IL   
...        ...         ...             ...   ...   
1856     90755       75202          Dallas    TX   
1857     94610       84627         Ephraim    UT   
1858     84452       60301        Oak Park    IL   
1859     94629       84647  Mount Pleasant    UT   
1860     91695       77046         Houston    TX   

                                 Metro       CountyName  SizeRank  2010-09  \
0          New York-Newark-Jersey City  New York County         1   2930.0   
1             Chicago-Naperville-Elgin      Cook County         2   1447.0   
2          New York-Newark-Jersey City  New York County         3   2797.0   
3     Houston-The Woodlands-Sugar Land    Harris County         4      NaN   
4             Chicago-Naperville-Elgin      Cook County         5   1437.0   
...                                ...              ...       ...      ...   
1856       Dallas-Fort Worth-Arlington    Dallas County      1857    997.0   
1857                               NaN   Sanpete County      1858      NaN   
1858          Chicago-Naperville-Elgin      Cook County      1859      NaN   
1859                               NaN   Sanpete County      1860      NaN   
1860  Houston-The Woodlands-Sugar Land    Harris County      1861      NaN   

      2010-10  2010-11  ...  2019-04  2019-05  2019-06  2019-07  2019-08  \
0      2952.0   2926.0  ...   3484.0   3523.0   3573.0   3622.0   3664.0   
1      1465.0   1469.0  ...   1731.0   1749.0   1768.0   1787.0   1801.0   
2      2811.0   2813.0  ...   3449.0   3459.0   3479.0   3491.0   3510.0   
3         NaN      NaN  ...   1226.0   1250.0   1264.0   1251.0   1246.0   
4      1469.0   1490.0  ...   1888.0   1911.0   1934.0   1950.0   1953.0   
...       ...      ...  ...      ...      ...      ...      ...      ...   
1856   1019.0   1042.0  ...   1615.0   1619.0   1628.0      NaN      NaN   
1857      NaN      NaN  ...      NaN      NaN      NaN      NaN      NaN   
1858      NaN      NaN  ...   1447.0   1483.0   1495.0   1511.0   1537.0   
1859      NaN      NaN  ...      NaN      NaN      NaN      NaN      NaN   
1860      NaN      NaN  ...      NaN      NaN      NaN      NaN      NaN   

      2019-09  2019-10  2019-11  2019-12  2020-01  
0      3698.0   3704.0   3692.0   3715.0   3676.0  
1      1806.0   1789.0   1761.0   1747.0   1731.0  
2      3521.0   3574.0   3608.0   3561.0      NaN  
3      1252.0   1280.0   1310.0   1294.0   1286.0  
4      1952.0   1921.0   1958.0      NaN   1916.0  
...       ...      ...      ...      ...      ...  
1856   1729.0   1697.0   1739.0      NaN      NaN  
1857      NaN      NaN      NaN   1298.0      NaN  
1858   1570.0   1580.0   1625.0      NaN      NaN  
1859      NaN      NaN      NaN      NaN   1221.0  
1860      NaN      NaN      NaN      NaN      NaN  

[1861 rows x 120 columns]

In [204]:
# Huge number of rows with tons of missing values.
sorted(np.sum(zillow.isnull(), axis=1), reverse=True)

[113,
 113,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 112,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 111,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 110,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 109,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 108,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 107,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 106,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105,
 105

In [205]:
np.mean(np.sum(zillow.isnull(), axis=1) > 55)

0.23589468027941968

In [206]:
np.mean(np.sum(zillow.isnull(), axis=1) > 35)

0.5303600214938206

I did not realize the multifam data has so many missing values. half of all price rows are missing 1/3 of their values, and a quarter of all rows are missing half of their values. I really really don't know that we should use this data. Or at the very least, we should be dropping a HUGE amount of it. My interpolate function was built using the ZORI data.

In [207]:
# Well, we don't have any use for data before our first census dataset. 
# That is 2011, means sept 2012, which is a forecast for sept 2015. 
# Should delete all data older than sept 2015 and then look at missingness.

usable_date_index = list(zillow.columns).index('2015-09')
first_date_index = 7

zillow_trim = pd.merge(zillow.iloc[:,:7], zillow.iloc[:,63:], how='left', left_index=True, right_index=True)


In [208]:
np.mean(np.sum(zillow_trim.isnull(), axis=1) > 28)

0.16980118216012896

In [209]:
np.mean(np.sum(zillow_trim.isnull(), axis=1) > 18)

0.19881783987103707

Okay, slightly more manageable. 17% of rows still have half or more values missing, but only 20% of rows have 1/3 or more value missing. I think we might want to drop these 17% of rows that have 50%+ missing rent values. 

In [210]:
# Rows to drop:
zillow_trim[np.sum(zillow_trim.isnull(), axis=1) > 28]

RegionID  RegionName            City State  \
5        91940       77449            Katy    TX   
15       74101       37013       Nashville    TN   
25       92036       77573     League City    TX   
33       71067       30349       Riverdale    GA   
67       91889       77379          Spring    TX   
...        ...         ...             ...   ...   
1845     67594       23219        Richmond    VA   
1851     90778       75226          Dallas    TX   
1857     94610       84627         Ephraim    UT   
1859     94629       84647  Mount Pleasant    UT   
1860     91695       77046         Houston    TX   

                                           Metro        CountyName  SizeRank  \
5               Houston-The Woodlands-Sugar Land     Harris County         6   
15    Nashville-Davidson--Murfreesboro--Franklin   Davidson County        16   
25              Houston-The Woodlands-Sugar Land  Galveston County        26   
33                 Atlanta-Sandy Springs-Roswell    Clayton County        34   
67              Houston-The Woodlands-Sugar Land     Harris County        68   
...                                          ...               ...       ...   
1845                                    Richmond     Richmond City      1846   
1851                 Dallas-Fort Worth-Arlington     Dallas County      1852   
1857                                         NaN    Sanpete County      1858   
1859                                         NaN    Sanpete County      1860   
1860            Houston-The Woodlands-Sugar Land     Harris County      1861   

      2015-05  2015-06  2015-07  ...  2019-04  2019-05  2019-06  2019-07  \
5         NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
15        NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
25        NaN      NaN      NaN  ...   1157.0   1150.0   1131.0   1111.0   
33        NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
67        NaN      NaN      NaN  ...   1245.0   1238.0   1231.0   1247.0   
...       ...      ...      ...  ...      ...      ...      ...      ...   
1845      NaN      NaN      NaN  ...      NaN   1207.0   1210.0   1211.0   
1851      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
1857      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
1859      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   
1860      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   

      2019-08  2019-09  2019-10  2019-11  2019-12  2020-01  
5         NaN      NaN   1314.0   1321.0      NaN   1353.0  
15        NaN      NaN   1028.0   1015.0   1013.0   1010.0  
25     1100.0   1092.0   1103.0   1098.0   1094.0   1110.0  
33        NaN   1120.0   1101.0   1087.0   1057.0      NaN  
67     1262.0   1296.0   1292.0      NaN      NaN      NaN  
...       ...      ...      ...      ...      ...      ...  
1845   1210.0   1212.0   1237.0   1242.0   1269.0   1286.0  
1851      NaN   1364.0      NaN   1413.0   1435.0   1436.0  
1857      NaN      NaN      NaN      NaN   1298.0      NaN  
1859      NaN      NaN      NaN      NaN      NaN   1221.0  
1860      NaN      NaN      NaN      NaN      NaN      NaN  

[316 rows x 64 columns]

In [211]:
# Rows to keep:
zillow_trim = zillow_trim[np.sum(zillow_trim.isnull(), axis=1) < 28]
zillow_trim

RegionID  RegionName         City State  \
0        61639       10025     New York    NY   
1        84654       60657      Chicago    IL   
2        61637       10023     New York    NY   
3        91982       77494         Katy    TX   
4        84616       60614      Chicago    IL   
...        ...         ...          ...   ...   
1853     92311       78215  San Antonio    TX   
1854     58656        2142    Cambridge    MA   
1855     65393       18232     Lansford    PA   
1856     90755       75202       Dallas    TX   
1858     84452       60301     Oak Park    IL   

                                 Metro        CountyName  SizeRank  2015-05  \
0          New York-Newark-Jersey City   New York County         1   3620.0   
1             Chicago-Naperville-Elgin       Cook County         2   1687.0   
2          New York-Newark-Jersey City   New York County         3   3469.0   
3     Houston-The Woodlands-Sugar Land     Harris County         4   1214.0   
4             Chicago-Naperville-Elgin       Cook County         5   1785.0   
...                                ...               ...       ...      ...   
1853         San Antonio-New Braunfels      Bexar County      1854      NaN   
1854           Boston-Cambridge-Newton  Middlesex County      1855   3174.0   
1855        Allentown-Bethlehem-Easton     Carbon County      1856    727.0   
1856       Dallas-Fort Worth-Arlington     Dallas County      1857   1457.0   
1858          Chicago-Naperville-Elgin       Cook County      1859      NaN   

      2015-06  2015-07  ...  2019-04  2019-05  2019-06  2019-07  2019-08  \
0      3655.0   3685.0  ...   3484.0   3523.0   3573.0   3622.0   3664.0   
1      1703.0   1728.0  ...   1731.0   1749.0   1768.0   1787.0   1801.0   
2      3508.0   3560.0  ...   3449.0   3459.0   3479.0   3491.0   3510.0   
3      1214.0   1242.0  ...   1226.0   1250.0   1264.0   1251.0   1246.0   
4      1799.0   1808.0  ...   1888.0   1911.0   1934.0   1950.0   1953.0   
...       ...      ...  ...      ...      ...      ...      ...      ...   
1853      NaN      NaN  ...   1268.0      NaN   1334.0   1304.0      NaN   
1854   3125.0   3137.0  ...   2734.0   2720.0   2745.0   2777.0   2811.0   
1855    740.0    757.0  ...    781.0    773.0    769.0    762.0    758.0   
1856   1457.0   1460.0  ...   1615.0   1619.0   1628.0      NaN      NaN   
1858      NaN      NaN  ...   1447.0   1483.0   1495.0   1511.0   1537.0   

      2019-09  2019-10  2019-11  2019-12  2020-01  
0      3698.0   3704.0   3692.0   3715.0   3676.0  
1      1806.0   1789.0   1761.0   1747.0   1731.0  
2      3521.0   3574.0   3608.0   3561.0      NaN  
3      1252.0   1280.0   1310.0   1294.0   1286.0  
4      1952.0   1921.0   1958.0      NaN   1916.0  
...       ...      ...      ...      ...      ...  
1853      NaN   1194.0   1188.0   1186.0      NaN  
1854   2821.0   2799.0      NaN   3102.0      NaN  
1855    758.0      NaN      NaN      NaN      NaN  
1856   1729.0   1697.0   1739.0      NaN      NaN  
1858   1570.0   1580.0   1625.0      NaN      NaN  

[1545 rows x 64 columns]

In [212]:
# Total number of NaN's in trimmed data
zillow_trim.isnull().sum().sum()

4394

In [213]:
# Fill interior NaN's
zillow_trim = pd.merge(zillow_trim.iloc[:,0:7], zillow_trim.iloc[:,8:].interpolate(limit_area='inside', axis=1), left_index=True, right_index=True)

In [214]:
# Total number of beginning / end NaN's
zillow_trim.isnull().sum().sum()

1787

In [215]:
#zillow_trim.to_csv('temp_zillow_prefill.csv')

- 2105-06 to 2020-01 time frame (need sept for each year).
- Dropped 316 of 1860 rows that had 50%+ missing values
- Initial NaN count for this time frame post-drop: 4,394 / 86,464
- Filled inside missing values using linear interpolation, missing count after: 1,787



In [216]:
zillow_trim = zillow_trim.reset_index(drop=True)

In [217]:
zillow_trim.iloc[-50:-20, 6:]

SizeRank  2015-06  2015-07      2015-08      2015-09      2015-10  \
1495      1795   1185.0   1192.0  1203.000000  1208.000000  1215.000000   
1496      1796   1163.0   1179.0  1194.000000  1207.000000  1215.000000   
1497      1797    888.0    882.0   877.000000   881.000000   889.000000   
1498      1798   1170.0   1181.0  1195.000000  1200.000000  1198.000000   
1499      1799   3190.0   3155.0  3125.000000  3095.000000  3053.000000   
1500      1801    701.0    704.0   709.000000   717.000000   720.000000   
1501      1802   1852.0   1858.0  1876.000000  1897.000000  1905.000000   
1502      1804   1218.0   1218.0  1221.000000  1236.000000  1251.000000   
1503      1805   1006.0   1004.0  1020.000000  1035.000000  1066.000000   
1504      1807    673.0    690.0   707.000000   716.000000   713.000000   
1505      1808   2449.0   2454.0  2465.000000  2479.000000  2482.000000   
1506      1810   3871.0   3912.0  3948.000000  4063.000000  4178.000000   
1507      1812   2693.0   2703.0  2663.000000  2642.000000  2647.000000   
1508      1814      NaN      NaN          NaN          NaN          NaN   
1509      1816   2357.0   2383.0  2398.000000  2401.000000  2399.000000   
1510      1818   1866.0   1882.0  1898.000000  1915.000000  1935.000000   
1511      1819   1248.0   1257.0  1263.000000  1275.000000  1287.000000   
1512      1820   1773.0   1782.0  1802.000000  1834.000000  1858.000000   
1513      1821    970.0    972.0   969.000000   957.000000   960.000000   
1514      1822   2588.0   2593.0  2589.000000  2599.000000  2597.000000   
1515      1823    942.0    941.0   936.000000   942.000000   969.000000   
1516      1824      NaN      NaN          NaN          NaN          NaN   
1517      1825    750.0    764.0   768.000000   771.000000   767.000000   
1518      1826   2022.0   2045.0  2083.000000  2129.000000  2138.000000   
1519      1827   1059.0   1049.0  1048.000000  1040.000000  1058.000000   
1520      1828   2914.0   2942.0  2979.000000  3014.000000  3030.000000   
1521      1829      NaN   1349.0  1362.769231  1376.538462  1390.307692   
1522      1830   1033.0   1033.0  1030.000000  1027.000000  1035.000000   
1523      1833   1049.0   1056.0  1050.000000  1034.000000  1019.000000   
1524      1834    670.0    687.0   697.000000   699.000000   700.000000   

          2015-11      2015-12      2016-01      2016-02  ...      2019-04  \
1495  1226.000000  1227.000000  1229.000000  1217.000000  ...  1262.000000   
1496  1227.000000  1230.000000  1233.000000  1233.000000  ...  1211.000000   
1497   876.000000   872.000000   875.000000   883.000000  ...   881.428571   
1498  1213.000000  1220.000000  1216.000000  1206.000000  ...  1425.000000   
1499  3023.000000  3002.000000  3024.000000  3070.000000  ...  3339.000000   
1500   719.000000   715.000000   710.000000   707.000000  ...   735.000000   
1501  1900.000000  1922.000000  1934.000000  1933.000000  ...  2057.000000   
1502  1263.000000  1275.000000  1282.000000  1283.000000  ...  1465.000000   
1503  1070.000000  1065.000000  1058.000000  1058.000000  ...  1044.000000   
1504   701.000000   692.000000   686.000000   688.000000  ...   722.000000   
1505  2470.000000  2453.000000  2428.000000  2414.000000  ...  2523.000000   
1506  4293.000000  4408.000000  4523.000000  4638.000000  ...  3882.000000   
1507  2656.000000  2647.000000  2661.000000  2672.000000  ...  2918.000000   
1508  1914.000000  1970.500000  2027.000000  2074.500000  ...          NaN   
1509  2382.000000  2375.000000  2376.000000  2391.000000  ...  2529.000000   
1510  1944.000000  1957.000000  1964.000000  1960.000000  ...  2179.000000   
1511  1296.000000  1303.000000  1309.000000  1306.000000  ...  1295.000000   
1512  1871.000000  1878.000000  1874.000000  1877.000000  ...  2138.000000   
1513   976.000000   978.000000   976.000000   963.000000  ...  1047.000000   
1514  2587.000000  2552.000000  2536.000000  2530.000000  ...  2672.000000   
1515   975.000000   978.000000   973.00000

In [218]:
# Fill in the beginning NaN's. 
for row in range(len(zillow_trim)):
    i = 7 # first rent column
    while True:
        if pd.isnull(zillow_trim.iloc[row, i]):
            i+=1
        else:
            break
    i -= 1
    while i > 6:
        zillow_trim.iloc[row, i] = (zillow_trim.iloc[row, i+12] * .9615)
        i-=1

In [219]:
zillow_trim.iloc[-50:-20, 6:]

SizeRank      2015-06      2015-07      2015-08      2015-09  \
1495      1795  1185.000000  1192.000000  1203.000000  1208.000000   
1496      1796  1163.000000  1179.000000  1194.000000  1207.000000   
1497      1797   888.000000   882.000000   877.000000   881.000000   
1498      1798  1170.000000  1181.000000  1195.000000  1200.000000   
1499      1799  3190.000000  3155.000000  3125.000000  3095.000000   
1500      1801   701.000000   704.000000   709.000000   717.000000   
1501      1802  1852.000000  1858.000000  1876.000000  1897.000000   
1502      1804  1218.000000  1218.000000  1221.000000  1236.000000   
1503      1805  1006.000000  1004.000000  1020.000000  1035.000000   
1504      1807   673.000000   690.000000   707.000000   716.000000   
1505      1808  2449.000000  2454.000000  2465.000000  2479.000000   
1506      1810  3871.000000  3912.000000  3948.000000  4063.000000   
1507      1812  2693.000000  2703.000000  2663.000000  2642.000000   
1508      1814  2047.995000  2070.109500  2001.843000  1933.576500   
1509      1816  2357.000000  2383.000000  2398.000000  2401.000000   
1510      1818  1866.000000  1882.000000  1898.000000  1915.000000   
1511      1819  1248.000000  1257.000000  1263.000000  1275.000000   
1512      1820  1773.000000  1782.000000  1802.000000  1834.000000   
1513      1821   970.000000   972.000000   969.000000   957.000000   
1514      1822  2588.000000  2593.000000  2589.000000  2599.000000   
1515      1823   942.000000   941.000000   936.000000   942.000000   
1516      1824  1079.644313  1095.869625  1112.094938  1128.320250   
1517      1825   750.000000   764.000000   768.000000   771.000000   
1518      1826  2022.000000  2045.000000  2083.000000  2129.000000   
1519      1827  1059.000000  1049.000000  1048.000000  1040.000000   
1520      1828  2914.000000  2942.000000  2979.000000  3014.000000   
1521      1829  1442.693769  1349.000000  1362.769231  1376.538462   
1522      1830  1033.000000  1033.000000  1030.000000  1027.000000   
1523      1833  1049.000000  1056.000000  1050.000000  1034.000000   
1524      1834   670.000000   687.000000   697.000000   699.000000   

          2015-10      2015-11      2015-12      2016-01      2016-02  ...  \
1495  1215.000000  1226.000000  1227.000000  1229.000000  1217.000000  ...   
1496  1215.000000  1227.000000  1230.000000  1233.000000  1233.000000  ...   
1497   889.000000   876.000000   872.000000   875.000000   883.000000  ...   
1498  1198.000000  1213.000000  1220.000000  1216.000000  1206.000000  ...   
1499  3053.000000  3023.000000  3002.000000  3024.000000  3070.000000  ...   
1500   720.000000   719.000000   715.000000   710.000000   707.000000  ...   
1501  1905.000000  1900.000000  1922.000000  1934.000000  1933.000000  ...   
1502  1251.000000  1263.000000  1275.000000  1282.000000  1283.000000  ...   
1503  1066.000000  1070.000000  1065.000000  1058.000000  1058.000000  ...   
1504   713.000000   701.000000   692.000000   686.000000   688.000000  ...   
1505  2482.000000  2470.000000  2453.000000  2428.000000  2414.000000  ...   
1506  4178.000000  4293.000000  4408.000000  4523.000000  4638.000000  ...   
1507  2647.000000  2656.000000  2647.000000  2661.000000  2672.000000  ...   
1508  1894.155000  1914.000000  1970.500000  2027.000000  2074.500000  ...   
1509  2399.000000  2382.000000  2375.000000  2376.000000  2391.000000  ...   
1510  1935.000000  1944.000000  1957.000000  1964.000000  1960.000000  ...   
1511  1287.000000  1296.000000  1303.000000  1309.000000  1306.000000  ...   
1512  1858.000000  1871.000000  1878.000000  1874.000000  1877.000000  ...   
1513   960.000000   976.000000   978.000000   976.000000   963.000000  ...   
1514  2597.000000  2587.000000  2552.000000  2536.000000  2530.000000  ...   
1515   969.000000   975.000000   978.000000   973.000000   993.000000  ...   
1516  1144.545562  1160.770875  1176.996188  1193.221500  1047.000000  ...   
1517   767.000000   757.000000   747.0000

In [220]:
zillow_trim.iloc[-20:, -10:]

2019-04  2019-05  2019-06      2019-07      2019-08      2019-09  \
1525   2797.0   2853.0   2911.0  2959.000000  2989.000000  2995.000000   
1526    899.0    894.0    897.0   905.000000   913.000000   918.000000   
1527   1423.0   1414.0   1410.0  1409.000000  1404.000000  1408.000000   
1528      NaN      NaN      NaN          NaN          NaN          NaN   
1529   1111.0   1113.0   1123.0  1130.000000  1138.000000  1141.000000   
1530    871.0    854.0    841.0   821.000000   811.000000   808.000000   
1531   2592.0   2655.0   2727.0  2787.000000  2829.000000  2850.000000   
1532    807.0    811.0    815.0   820.000000   824.000000   829.000000   
1533   1922.0   1930.0   1939.0  1946.000000  1954.000000  1958.000000   
1534   2968.0   2979.0   3001.0  3035.000000  3062.000000  3085.000000   
1535   1864.0   1866.0   1866.0  1880.000000  1894.000000  1905.000000   
1536   1329.0   1329.0   1331.0  1339.000000  1341.000000  1342.000000   
1537   1092.0   1111.0   1108.0  1105.000000  1113.000000  1113.000000   
1538   2145.0   2155.0   2166.0  2183.000000  2206.000000  2228.000000   
1539    937.0    949.0    964.0   974.000000   978.000000   978.000000   
1540   1268.0   1301.0   1334.0  1304.000000  1267.333333  1230.666667   
1541   2734.0   2720.0   2745.0  2777.000000  2811.000000  2821.000000   
1542    781.0    773.0    769.0   762.000000   758.000000   758.000000   
1543   1615.0   1619.0   1628.0  1661.666667  1695.333333  1729.000000   
1544   1447.0   1483.0   1495.0  1511.000000  1537.000000  1570.000000   

          2019-10      2019-11  2019-12  2020-01  
1525  2922.000000  2849.000000   2832.0   2849.0  
1526   924.000000   922.000000    925.0    927.0  
1527  1434.000000  1433.000000   1423.0   1425.0  
1528          NaN          NaN      NaN      NaN  
1529  1170.000000  1157.500000   1145.0      NaN  
1530   809.000000   801.000000    793.0    806.0  
1531  2820.000000  2828.000000   2773.0   2847.0  
1532   844.000000   847.000000    850.0      NaN  
1533  1918.000000  1938.000000   1920.0   1918.0  
1534  3049.000000  3054.000000   3019.0   2992.0  
1535  1906.000000  1901.000000   1902.0   1847.0  
1536  1343.666667  1345.333333   1347.0   1350.0  
1537  1146.000000          NaN      NaN      NaN  
1538  2217.000000  2230.000000   2200.0   2192.0  
1539   998.333333  1018.666667   1039.0   1047.0  
1540  1194.000000  1188.000000   1186.0      NaN  
1541  2799.000000  2950.500000   3102.0      NaN  
1542          NaN          NaN      NaN      NaN  
1543  1697.000000  1739.000000      NaN      NaN  
1544  1580.000000  1625.000000      NaN      NaN

In [221]:
# Fill in the ending NaN's. 
for row in range(len(zillow_trim)):
    i = 62 # last rent column
    while True:
        if pd.isnull(zillow_trim.iloc[row, i]):
            i-=1
        else:
            break
    i += 1
    while i < 63:
        zillow_trim.iloc[row, i] = (zillow_trim.iloc[row, i-12] * 1.04)
        i+=1

In [222]:
zillow_trim.iloc[-20:, -10:]

2019-04  2019-05  2019-06      2019-07      2019-08      2019-09  \
1525  2797.00  2853.00  2911.00  2959.000000  2989.000000  2995.000000   
1526   899.00   894.00   897.00   905.000000   913.000000   918.000000   
1527  1423.00  1414.00  1410.00  1409.000000  1404.000000  1408.000000   
1528  1457.04  1499.68  1521.52  1526.720000  1549.600000  1571.440000   
1529  1111.00  1113.00  1123.00  1130.000000  1138.000000  1141.000000   
1530   871.00   854.00   841.00   821.000000   811.000000   808.000000   
1531  2592.00  2655.00  2727.00  2787.000000  2829.000000  2850.000000   
1532   807.00   811.00   815.00   820.000000   824.000000   829.000000   
1533  1922.00  1930.00  1939.00  1946.000000  1954.000000  1958.000000   
1534  2968.00  2979.00  3001.00  3035.000000  3062.000000  3085.000000   
1535  1864.00  1866.00  1866.00  1880.000000  1894.000000  1905.000000   
1536  1329.00  1329.00  1331.00  1339.000000  1341.000000  1342.000000   
1537  1092.00  1111.00  1108.00  1105.000000  1113.000000  1113.000000   
1538  2145.00  2155.00  2166.00  2183.000000  2206.000000  2228.000000   
1539   937.00   949.00   964.00   974.000000   978.000000   978.000000   
1540  1268.00  1301.00  1334.00  1304.000000  1267.333333  1230.666667   
1541  2734.00  2720.00  2745.00  2777.000000  2811.000000  2821.000000   
1542   781.00   773.00   769.00   762.000000   758.000000   758.000000   
1543  1615.00  1619.00  1628.00  1661.666667  1695.333333  1729.000000   
1544  1447.00  1483.00  1495.00  1511.000000  1537.000000  1570.000000   

          2019-10      2019-11  2019-12      2020-01  
1525  2922.000000  2849.000000  2832.00  2849.000000  
1526   924.000000   922.000000   925.00   927.000000  
1527  1434.000000  1433.000000  1423.00  1425.000000  
1528  1588.080000  1592.240000  1602.64  1503.424000  
1529  1170.000000  1157.500000  1145.00  1154.400000  
1530   809.000000   801.000000   793.00   806.000000  
1531  2820.000000  2828.000000  2773.00  2847.000000  
1532   844.000000   847.000000   850.00   836.160000  
1533  1918.000000  1938.000000  1920.00  1918.000000  
1534  3049.000000  3054.000000  3019.00  2992.000000  
1535  1906.000000  1901.000000  1902.00  1847.000000  
1536  1343.666667  1345.333333  1347.00  1350.000000  
1537  1146.000000  1043.120000  1024.40  1056.640000  
1538  2217.000000  2230.000000  2200.00  2192.000000  
1539   998.333333  1018.666667  1039.00  1047.000000  
1540  1194.000000  1188.000000  1186.00  1318.720000  
1541  2799.000000  2950.500000  3102.00  3020.680000  
1542   813.280000   813.280000   815.36   835.120000  
1543  1697.000000  1739.000000  1669.20  1671.626667  
1544  1580.000000  1625.000000  1476.80  1462.240000

In [223]:
zillow_trim

RegionID  RegionName         City State  \
0        61639       10025     New York    NY   
1        84654       60657      Chicago    IL   
2        61637       10023     New York    NY   
3        91982       77494         Katy    TX   
4        84616       60614      Chicago    IL   
...        ...         ...          ...   ...   
1540     92311       78215  San Antonio    TX   
1541     58656        2142    Cambridge    MA   
1542     65393       18232     Lansford    PA   
1543     90755       75202       Dallas    TX   
1544     84452       60301     Oak Park    IL   

                                 Metro        CountyName  SizeRank  \
0          New York-Newark-Jersey City   New York County         1   
1             Chicago-Naperville-Elgin       Cook County         2   
2          New York-Newark-Jersey City   New York County         3   
3     Houston-The Woodlands-Sugar Land     Harris County         4   
4             Chicago-Naperville-Elgin       Cook County         5   
...                                ...               ...       ...   
1540         San Antonio-New Braunfels      Bexar County      1854   
1541           Boston-Cambridge-Newton  Middlesex County      1855   
1542        Allentown-Bethlehem-Easton     Carbon County      1856   
1543       Dallas-Fort Worth-Arlington     Dallas County      1857   
1544          Chicago-Naperville-Elgin       Cook County      1859   

          2015-06      2015-07      2015-08  ...  2019-04  2019-05  2019-06  \
0     3655.000000  3685.000000  3673.000000  ...   3484.0   3523.0   3573.0   
1     1703.000000  1728.000000  1746.000000  ...   1731.0   1749.0   1768.0   
2     3508.000000  3560.000000  3596.000000  ...   3449.0   3459.0   3479.0   
3     1214.000000  1242.000000  1276.000000  ...   1226.0   1250.0   1264.0   
4     1799.000000  1808.000000  1817.000000  ...   1888.0   1911.0   1934.0   
...           ...          ...          ...  ...      ...      ...      ...   
1540  1178.714869  1208.298301  1212.920712  ...   1268.0   1301.0   1334.0   
1541  3125.000000  3137.000000  3185.000000  ...   2734.0   2720.0   2745.0   
1542   740.000000   757.000000   776.000000  ...    781.0    773.0    769.0   
1543  1457.000000  1460.000000  1467.000000  ...   1615.0   1619.0   1628.0   
1544  1350.668567  1387.647857  1392.270268  ...   1447.0   1483.0   1495.0   

          2019-07      2019-08      2019-09  2019-10  2019-11  2019-12  \
0     3622.000000  3664.000000  3698.000000  3704.00  3692.00  3715.00   
1     1787.000000  1801.000000  1806.000000  1789.00  1761.00  1747.00   
2     3491.000000  3510.000000  3521.000000  3574.00  3608.00  3561.00   
3     1251.000000  1246.000000  1252.000000  1280.00  1310.00  1294.00   
4     1950.000000  1953.000000  1952.000000  1921.00  1958.00  1937.00   
...           ...          ...          ...      ...      ...      ...   
1540  1304.000000  1267.333333  1230.666667  1194.00  1188.00  1186.00   
1541  2777.000000  2811.000000  2821.000000  2799.00  2950.50  3102.00   
1542   762.000000   758.000000   758.000000   813.28   813.28   815.36   
1543  1661.666667  1695.333333  1729.000000  1697.00  1739.00  1669.20   
1544  1511.000000  1537.000000  1570.000000  1580.00  1625.00  1476.80   

          2020-01  
0     3676.000000  
1     1731.000000  
2     3655.600000  
3     1286.000000  
4     1916.000000  
...           ...  
1540  1318.720000  
1541  3020.680000  
1542   835.120000  
1543  1671.626667  
1544  1462.240000  

[1545 rows x 63 columns]

In [224]:
zillow_trim.to_csv('Zillow_ZRI_Multifam_Trimmed_Filled.csv')